metrics/cloudwatch/cloudwatch.go

```diff
@@ -12,81 +13,131 @@ import (
 	"github.com/go-kit/kit/log"
 	"github.com/go-kit/kit/metrics"
 	"github.com/go-kit/kit/metrics/generic"
+	"github.com/go-kit/kit/metrics/internal/lv"
 )
 
 const (
 	maxConcurrentRequests = 20
 )
 
+type Percentiles []struct {
+	s string
+	f float64
+}
+
 // CloudWatch receives metrics observations and forwards them to CloudWatch.
 // Create a CloudWatch object, use it to create metrics, and pass those metrics as
 // dependencies to the components that will use them.
 //
 // To regularly report metrics to CloudWatch, use the WriteLoop helper method.
 type CloudWatch struct {
-	mtx                   sync.RWMutex
-	sem                   chan struct{}
-	namespace             string
-	numConcurrentRequests int
-	svc                   cloudwatchiface.CloudWatchAPI
-	counters              map[string]*counter
-	gauges                map[string]*gauge
-	histograms            map[string]*histogram
+	mtx        sync.RWMutex
+	sem        chan struct{}
+	namespace  string
+	svc        cloudwatchiface.CloudWatchAPI
+	counters   *lv.Space
+	gauges     *lv.Space
+	histograms *lv.Space
+	*cwoptions
+}
+
+type cwoptions struct {
+	percentiles           Percentiles
 	logger                log.Logger
+	numConcurrentRequests int
+}
+
+type option func(*cwoptions)
+
+func (s *cwoptions) apply(opt option) {
+	if opt != nil {
+		opt(s)
+	}
+}
+
+func WithLogger(logger log.Logger) option {
+	return func(o *cwoptions) {
+		o.logger = logger
+	}
+}
+
+func WithPercentiles(p Percentiles) option {
+	return func(o *cwoptions) {
+		validated := Percentiles{}
+		for _, entry := range p {
+			if entry.f < 0 || entry.f > 1 {
+				continue // illegal entry
+			}
+			validated = append(validated, entry)
+		}
+		o.percentiles = validated
+	}
+}
```

Comment:
No other metrics package has user-modifiable percentiles. Why is it important here?